calico_upgrade/pkg/migrate/felixconfig.go

```diff
@@ -0,0 +1,279 @@
+			return err
+		}
+	}
+
+	return nil
+}
+
+// This function converts a slice of v1 KVPairs into the appropriate v3 values and
+// merges the results into a single v3 resource Spec for felix configuration (global
+// or per host) or a clusterInfo.
+// Conversion errors are added to the ConvertedData struct.
+func (fc *felixConfig) parseFelixConfigV1IntoResourceV3(
+	kvps []*model.KVPair,
+	res conversionv1v3.Resource,
+	data *ConvertedData,
+) error {
+	logCxtRes := log.WithFields(log.Fields{
+		"kind": res.GetObjectKind().GroupVersionKind().Kind,
+		"name": res.GetObjectMeta().GetName(),
+	})
+
+	// Convert the KVP slice into a name value map.
+	config := map[string]string{}
+	for _, kvp := range kvps {
+		if kvp.Value == nil {
+			continue
+		}
+		switch key := kvp.Key.(type) {
+		case model.GlobalConfigKey:
+			config[key.Name] = kvp.Value.(string)
+		case model.HostConfigKey:
+			config[key.Name] = kvp.Value.(string)
+		}
+	}
+
+	// Extract the Spec from the resource FelixConfiguration or ClusterInfo.
+	specValue := reflect.ValueOf(res).Elem().FieldByName("Spec")
+	if !specValue.IsValid() {
+		return fmt.Errorf("unable to process config resource type: %v", res)
+	}
+
+	// Loop through the Spec setting each field from the supplied KVPair data.
+	setField := false
+	specType := specValue.Type()
+	for i := 0; i < specType.NumField(); i++ {
+		field := specType.Field(i)
+		fieldValue := specValue.Field(i)
+
+		// Get the v1 config value associated with the field.
+		configName := fc.getConfigName(field)
+		logCxt := logCxtRes.WithFields(log.Fields{
+			"field":  field.Name,
+			"config": configName,
+		})
+		configStrValue, ok := config[configName]
+		if !ok {
+			logCxt.Debug("config value is not configured in v1")
+			continue
+		}
+
+		isPtr := field.Type.Kind() == reflect.Ptr
+		fieldName := field.Name
+
+		switch {
+		case strings.HasPrefix(fieldName, "Failsafe"):
+			// Special-case the Failsafe ports - these require parsing and settings as a struct.
+			if configStrValue == "none" {
+				// Has no failsafe ports
+				vProtoPort := &[]apiv3.ProtoPort{}
+				fieldValue.Set(reflect.ValueOf(vProtoPort))
+				continue
+			}
+
+			vProtoPort, err := fc.parseProtoPort(configStrValue)
+			if err != nil {
+				data.ConversionErrors = append(data.ConversionErrors, ConversionError{})
+			}
+			fieldValue.Set(reflect.ValueOf(vProtoPort)) // pointer to proto port slice.
+			continue
+		case strings.HasPrefix(fieldName, "LogSeverity"):
+			// The log level fields need to have their value converted to the appropriate v3 value,
+			// but other than that are treated as normal string fields.
+			configStrValue = convertLogLevel(configStrValue)
+		}
+
+		// Set the field value based on the field type.
+		var kind reflect.Kind
+		if isPtr {
+			kind = field.Type.Elem().Kind()
+		} else {
+			kind = fieldValue.Kind()
+		}
+
+		switch kind {
+		case reflect.Uint32:
+			if value, err := strconv.ParseUint(configStrValue, 10, 32); err != nil {
+				continue
```

Comment:
is there value in debug logging these ParseUint failures?

Reply:
Yes - i'll log.

Reply:
omg - good spot - this was supposed to add entries to the ConversionError.